library/x509write_crt.c

```diff
@@ -167,66 +172,87 @@ int mbedtls_x509write_crt_set_basic_constraints( mbedtls_x509write_cert *ctx,
     unsigned char *c = buf + sizeof(buf);
     size_t len = 0;
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    size_t hash_length;
+#endif /* MBEDTLS_USE_PSA_CRYPTO */
 
     memset( buf, 0, sizeof(buf) );
-    MBEDTLS_ASN1_CHK_ADD( len,
-                mbedtls_pk_write_pubkey( &c, buf, ctx->subject_key ) );
-
+    MBEDTLS_ASN1_CHK_ADD( len, mbedtls_pk_write_pubkey( &c,
+                                                        buf,
+                                                        is_ca ?
+                                                          ctx->issuer_key :
+                                                          ctx->subject_key ) );
+
+
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    status = psa_hash_compute( PSA_ALG_SHA_1,
+                               buf + sizeof(buf) - len,
+                               len,
+                               buf + sizeof(buf) - 20,
+                               20,
+                               &hash_length );
+    if( status != PSA_SUCCESS )
+    {
+        return( MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED );
+    }
+#else
     ret = mbedtls_sha1( buf + sizeof( buf ) - len, len,
-                            buf + sizeof( buf ) - 20 );
+                        buf + sizeof( buf ) - 20 );
     if( ret != 0 )
         return( ret );
+#endif /* MBEDTLS_USE_PSA_CRYPTO */
+
     c = buf + sizeof( buf ) - 20;
     len = 20;
 
     MBEDTLS_ASN1_CHK_ADD( len, mbedtls_asn1_write_len( &c, buf, len ) );
     MBEDTLS_ASN1_CHK_ADD( len,
-            mbedtls_asn1_write_tag( &c, buf, MBEDTLS_ASN1_OCTET_STRING ) );
+                          mbedtls_asn1_write_tag( &c, buf, tag ) );
 
-    return mbedtls_x509write_crt_set_extension( ctx,
-                 MBEDTLS_OID_SUBJECT_KEY_IDENTIFIER,
-                 MBEDTLS_OID_SIZE( MBEDTLS_OID_SUBJECT_KEY_IDENTIFIER ),
-                 0, buf + sizeof(buf) - len, len );
+    if( is_ca ) // writes AuthorityKeyIdentifier sequence
+    {
+        MBEDTLS_ASN1_CHK_ADD( len, mbedtls_asn1_write_len( &c, buf, len ));
+        MBEDTLS_ASN1_CHK_ADD( len,
+                              mbedtls_asn1_write_tag( &c,
+                                                      buf,
+                                                      MBEDTLS_ASN1_CONSTRUCTED |
+                                                      MBEDTLS_ASN1_SEQUENCE ) );
+    }
+    return mbedtls_x509write_crt_set_extension(
+                   ctx,
+                   is_ca ? MBEDTLS_OID_AUTHORITY_KEY_IDENTIFIER :
+                               MBEDTLS_OID_SUBJECT_KEY_IDENTIFIER,
+                   is_ca ? MBEDTLS_OID_SIZE(
+                               MBEDTLS_OID_AUTHORITY_KEY_IDENTIFIER ) :
+                           MBEDTLS_OID_SIZE(
+                               MBEDTLS_OID_SUBJECT_KEY_IDENTIFIER ),
+                   0,
+                   buf + sizeof( buf ) - len,
+                   len );
```

Comment:
This one is hard to read, perhalps this would be easier to read:
```
    if( is_ca )
        return mbedtls_x509write_crt_set_extension( ctx,
                     MBEDTLS_OID_AUTHORITY_KEY_IDENTIFIER,
                     MBEDTLS_OID_SIZE( MBEDTLS_OID_AUTHORITY_KEY_IDENTIFIER ),
                     0, buf + sizeof(buf) - len, len );

    return mbedtls_x509write_crt_set_extension( ctx,
                 MBEDTLS_OID_SUBJECT_KEY_IDENTIFIER,
                 MBEDTLS_OID_SIZE( MBEDTLS_OID_SUBJECT_KEY_IDENTIFIER ),
                 0, buf + sizeof(buf) - len, len );
```